sql/moz-fx-data-shared-prod/org_mozilla_ios_firefox/unified_metrics/query.py

```diff
@@ -0,0 +1,143 @@
+"""Generate a query for unifying legacy metrics and glean metrics.
+
+This approach is required because the order of columns is significant in order
+to be able to union two tables, even if the set of columns are the same."""
+
+import requests
+import json
+from bigquery_etl.format_sql.formatter import reformat
+from google.cloud import bigquery, exceptions
+import io
+
+
+def get_columns(schema):
+    """Return a list of columns corresponding to the schema.
+
+    Modified from https://github.com/mozilla-services/mozilla-pipeline-schemas/blob/c2dae92a7ed73e4774a897bc4d2a6a8608875339/mozilla_pipeline_schemas/utils.py#L30-L43
+    """
+
+    def traverse(prefix, columns):
+        res = []
+        for node in columns:
+            name = node["name"]
+            dtype = node["type"]
+            # we consider a repeated field a leaf node, and sorted for our purposes
+            if dtype == "RECORD" and node["mode"] != "REPEATED":
+                res += traverse(f"{prefix}.{name}", node["fields"])
+            else:
+                res += [f"{prefix}.{name} {dtype}"]
+        return res
+
+    res = traverse("root", schema)
+    return sorted(res)
+
+
+def generate_query(columns, table):
+    """Generate a SQL query given column names."""
+    # You would think that it's this simple (e.g. selecting columns), but it's
+    # not because we want to maintain the nested structure afterwards
+    #   formatted_columns = ",\n".join(columns)
+
+    acc = ""
+    prev = []
+    # sorting the columns is important
+    for col in sorted(columns):
+        split = col.split(".")
+        # check if we go deeper
+        if len(split) > 1 and len(split) > len(prev):
+            acc += "struct(" * (len(split) - len(prev))
+        # sometimes we have two structs that are the same depth e.g. metrics
+        if len(split) > 1 and len(split) == len(prev) and split[-2] != prev[-2]:
+            # ensure that we are not ending a struct with a comma
+            acc = acc.rstrip(",")
+            acc += f") as {prev[-2]},"
+            acc += "struct("
+        # pop out of the struct
+        if len(split) < len(prev):
+            diff = len(prev) - len(split)
+            # ignore the leaf
+            prev.pop()
+            for _ in range(diff):
+                c = prev.pop()
+                acc = acc.rstrip(",")
+                acc += f") as {c},"
+        acc += f"{col},"
+        prev = split
+    # clean up any columns
+    if len(prev) > 1:
+        prev.pop()
+        for c in reversed(prev):
+            acc = acc.rstrip(",")
+            acc += f") as {c},"
+
+    return reformat(f"select {acc} from `{table}`")
+
+
+def main():
+    # get the most schema deploy (to the nearest 15 minutes)
+    deploys_url = (
+        "https://protosaur.dev/mps-deploys/data/mozilla_pipeline_schemas/deploys.json"
```

Comment:
Hmm, view publishing will depend on this protosaur data being available?

Reply:
I'm hoping to be able to provide this information via dry run service instead: https://github.com/mozilla-services/cloudops-infra/pull/2916

Reply:
I guess we can directly figure out the commit hash by looking at the dataset tag via a bq api call.

Reply:
This snippet gets the commit-hash without going through a dryrun: 

```bash
bq = bigquery.Client()
commit_hash = (
    bq.get_dataset("moz-fx-data-shared-prod.telemetry")
    # labels are a dictionary on the dataset object
    .labels.get("schemas_build_id")
    # the schemas build id is the concatenation of a timestamp and commit hash
    .split("_")[1]
)
```